File: Hydra/RestoMenu.m
//
//  RestoMenu.m
//  Hydra
//
//  Created by Pieter De Baets on 17/07/12.
//  Copyright (c) 2012 Zeus WPI. All rights reserved.
//

#import "RestoMenu.h"

@implementation RestoMenu

- (NSString *)description
{
    NSUInteger count = [self.meat count] + [self.vegetables count];
    return [NSString stringWithFormat:@"<RestoMenu for %@ (%lu items) open=%@>",
                self.day, (unsigned long)count, NSStringFromBOOL(self.open)];
}

- (id)initWithCoder:(NSCoder *)coder
{
    if (self = [super init]) {
        self.day = [coder decodeObjectForKey:@"day"];
        self.open = [coder decodeBoolForKey:@"open"];
        self.meat = [coder decodeObjectForKey:@"meat"];
        self.vegetables = [coder decodeObjectForKey:@"vegetables"];
        self.soup = [coder decodeObjectForKey:@"soup"];
        self.lastUpdated = [coder decodeObjectForKey:@"lastUpdated"];
    }
    return self;
}

- (void)encodeWithCoder:(NSCoder *)coder
{
    [coder encodeObject:self.day forKey:@"day"];
    [coder encodeBool:self.open forKey:@"open"];
    [coder encodeObject:self.meat forKey:@"meat"];
    [coder encodeObject:self.vegetables forKey:@"vegetables"];
    [coder encodeObject:self.soup forKey:@"soup"];
    [coder encodeObject:self.lastUpdated forKey:@"lastUpdated"];
}

@end

@implementation RestoMenuItem

- (NSString *)description
{
    return [NSString stringWithFormat:@"<RestoMenuItem: %@ (%@)>", self.name, self.price];
}

- (id)initWithCoder:(NSCoder *)coder
{
    if (self = [super init]) {
        self.name = [coder decodeObjectForKey:@"name"];
        self.price = [coder decodeObjectForKey:@"price"];
        self.recommended = [coder decodeBoolForKey:@"recommended"];
    }
    return self;
}

- (void)encodeWithCoder:(NSCoder *)coder
{
    [coder encodeObject:self.name forKey:@"name"];
    [coder encodeObject:self.price forKey:@"price"];
    [coder encodeBool:self.recommended forKey:@"recommended"];
}

@end

File: Hydra/RestoMenu.swift
//
//  RestoMenu.swift
//  Hydra
//
//  Created by Feliciaan De Palmenaer on 02/03/2016.
//  Copyright © 2016 Zeus WPI. All rights reserved.
//

import Foundation

class RestoMenu: NSObject, Codable {

    var message: String?
    var date: Date
    var meals: [RestoMenuItem]?
    var open: Bool
    var vegetables: [String]?
    var lastUpdated: Date?

    var sideDishes: [RestoMenuItem]? {
        //TODO: cache...
        return meals?.filter({ $0.type == .Side })
    }

    var mainDishes: [RestoMenuItem]? {
        //TODO: cache...
        return meals?.filter({ $0.type != .Side})
    }

    init(message: String? = nil, date: Date, meals: [RestoMenuItem]? = nil, open: Bool = false, vegetables: [String]? = nil, lastUpdated: Date? = nil) {
        self.message = message
        self.date = date
        self.meals = meals
        self.open = open
        self.vegetables = vegetables
        self.lastUpdated = lastUpdated
    }

    //TODO: formatter.dateFormat = "yyyy-MM-dd"
    
    private enum CodingKeys: String, CodingKey {
        case message, date, open, meals, vegetables, lastUpdated
    }
}

class RestoMenuItem: NSObject, Codable {
    // TODO: fix with manual coding
    var kindVar: String?
    
    var kind: RestoMenuKind {
        get {
            if let kind = kindVar {
                if let k = RestoMenuKind(rawValue: kind) {
                    return k
                }
            }
            return .Other
        }
    }
    var name: String
    var price: String?
    var typeVar: String?
    
    var type: RestoMenuType {
        get {
            if let type = typeVar {
                if let t = RestoMenuType(rawValue: type) {
                    return t
                }
            }
            return .Other
        }
    }

    init(name: String, price: String? = nil, kind: RestoMenuKind = .Other, type: RestoMenuType = .Other) {
        self.name = name
        self.price = price
        self.kindVar = kind.rawValue
        self.typeVar = type.rawValue
    }

    fileprivate func restoMenuTypeFromString(_ type: String) -> RestoMenuType {
        var restoType = RestoMenuType.init(rawValue: type)
        if restoType == nil {
            restoType = .some(.Other)
        }
        return restoType!
    }

    fileprivate func restoMenuKindFromString(_ kind: String) -> RestoMenuKind {
        var restoKind = RestoMenuKind.init(rawValue: kind)
        if restoKind == nil {
            restoKind = .some(.Other)
        }
        return restoKind!
    }

    private enum CodingKeys: String, CodingKey {
        case name, price, typeVar = "type", kindVar = "kind"
    }
}

enum RestoMenuType: String, Codable {
    case Side = "side"
    case Main = "main"
    case Other = "other"
}

enum RestoMenuKind: String, Codable {
    case Soup = "soup"
    case Meat = "meat"
    case Vegetarian = "vegetarian"
    case Vegan = "vegan"
    case Fish = "fish"
    case Other = "other"
}
